metasrv/src/watcher/watcher_manager.rs

```diff
@@ -0,0 +1,242 @@
+//  Copyright 2021 Datafuse Labs.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+use core::ops::Range;
+use std::collections::BTreeMap;
+
+use common_base::tokio;
+use common_base::tokio::sync::mpsc;
+use common_base::tokio::sync::mpsc::Sender;
+use common_meta_raft_store::state_machine::StateMachineSubscriber;
+use common_meta_types::protobuf::watch_request::FilterType;
+use common_meta_types::protobuf::Event;
+use common_meta_types::protobuf::WatchRequest;
+use common_meta_types::protobuf::WatchResponse;
+use common_meta_types::SeqV;
+use common_range_set::RangeSet;
+use common_tracing::tracing;
+use tonic::Status;
+
+use super::WatcherKey;
+use super::WatcherStream;
+
+pub type WatcherId = i64;
+pub type WatcherStreamSender = Sender<Result<WatchResponse, Status>>;
+
+type CreateWatcherEvent = (WatchRequest, WatcherStreamSender);
+
+#[derive(Clone, Debug)]
+pub struct StateMachineKvData {
+    pub key: String,
+    pub prev: Option<SeqV>,
+    pub current: Option<SeqV>,
+}
+
+#[derive(Clone, Debug)]
+pub struct WatcherStateMachineSubscriber {
+    event_tx: mpsc::UnboundedSender<WatcherEvent>,
+}
+
+#[derive(Clone)]
+pub enum WatcherEvent {
+    CreateWatcherEvent(CreateWatcherEvent),
+    StateMachineKvDataEvent(StateMachineKvData),
+    ShutdownEvent(()),
+}
+
+#[derive(Debug)]
+pub struct WatcherManager {
+    event_tx: mpsc::UnboundedSender<WatcherEvent>,
+
+    pub subscriber: WatcherStateMachineSubscriber,
+}
+
+struct WatcherManagerCore {
+    event_rx: mpsc::UnboundedReceiver<WatcherEvent>,
+
+    watcher_streams: BTreeMap<WatcherId, WatcherStream>,
+
+    /// map range to WatcherId
+    watcher_range_set: RangeSet<String, WatcherKey>,
+
+    current_watcher_id: WatcherId,
+}
+
+impl WatcherManager {
+    pub fn create() -> Self {
+        let (event_tx, event_rx) = mpsc::unbounded_channel();
+
+        let core = WatcherManagerCore {
+            event_rx,
+            watcher_streams: BTreeMap::new(),
+            watcher_range_set: RangeSet::new(),
+            current_watcher_id: 1,
+        };
+
+        let _h = tokio::spawn(core.watcher_manager_main());
+
+        WatcherManager {
+            event_tx: event_tx.clone(),
+            subscriber: WatcherStateMachineSubscriber { event_tx },
+        }
+    }
+
+    pub fn stop(&self) {
+        let _ = self.event_tx.send(WatcherEvent::ShutdownEvent(()));
+    }
+
```

Comment:
Just dropping `self.event_tx` is quite enough to notify the core to shut down.

Then at the receiving end:
```rust
            tokio::select! {
                event = self.event_rx.recv() => {
                    if let Some(event) = event {...}
                    else { /* shutdown */ }
```